src/core/server/http/https_redirect_server.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Request, ResponseToolkit, Server } from 'hapi-latest';
+import { format as formatUrl } from 'url';
+
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+
+export class HttpsRedirectServer {
+  private server?: Server;
+
+  constructor(private readonly log: Logger) {}
+
+  public async start(config: HttpConfig) {
+    if (!config.ssl.enabled || config.ssl.redirectHttpFromPort === undefined) {
+      throw new Error(
+        'Redirect server cannot be started when [ssl.enabled] is set to `false`' +
+          ' or [ssl.redirectHttpFromPort] is not specified.'
+      );
+    }
+
+    this.log.info(
+      `starting HTTP --> HTTPS redirect server [${config.host}:${
+        config.ssl.redirectHttpFromPort
+      }]`
+    );
+
+    this.server = new Server({
```

Comment:
This server probably needs the fixes from https://github.com/elastic/kibana/commit/ec69a02d7e395a632a1591cdcb44466ffaff9f33 too, no?

Reply:
Since this request responds in the `onRequest` phase there isn't actually a need for `state`, `routes.cors` or `routes.payload` settings like we have elsewhere, but I'm slightly concerned about these three servers drifting in other ways down the road...

Reply:
> This server probably needs the fixes from ec69a02 too, no?

Theoretically yes, but not sure for what use cases redirect server it's used in practice. I'll add it here. That also means that we forgot to apply ec69a02 in master...

> Since this request responds in the onRequest phase there isn't actually a need for state, routes.cors or routes.payload settings like we have elsewhere, but I'm slightly concerned about these three servers drifting in other ways down the road...

Are you implying that we may want to use `configureHttpServer` (successor of `setup_connection`) here as well? Sounds like we can and probably even should do that.

Reply:
:heavy_check_mark:

Reply:
Nice, yeah, this looks good.